added Command and CommandParameter on CheckBox

pictos opened the pull request:
Description of Change

<!-- Enter description of the fix in this section -->

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #7394

pictos committed:
added Command and CommandParameter on CheckBox

pictos committed:
add unit test

pictos committed:
fix ICommandElement implementation

pictos committed:
code fix

pictos committed:
add public APIs

jsuarezruiz commented on src/Controls/tests/Core.UnitTests/CarouselViewTests.cs:
```diff
@@ -122,5 +122,37 @@ public void TestAddRemoveItems()
 		}
+
+		[Fact]
+		public void CheckBoxClickWhenCommandCanExecuteFalse()
```
This should be moved to CheckBoxUnitTests.

jsuarezruiz commented on src/Controls/src/Core/CheckBox/CheckBox.cs:
```diff
@@ -16,11 +18,40 @@ public partial class CheckBox : View, IElementConfiguration<CheckBox>, IBorderEl
 				}, defaultBindingMode: BindingMode.TwoWay);
 
+
+		public static readonly BindableProperty CommandProperty = BindableProperty.Create(nameof(Command), typeof(ICommand), typeof(CheckBox), null, propertyChanging: CommandElement.OnCommandChanging, propertyChanged: CommandElement.OnCommandChanged);
```
The new properties must be included in the PublicAPI files by platform. You can do it directly from Visual Studio but let me know if can help.

```
D:\a\_work\1\s\src\Controls\src\Core\CheckBox\CheckBox.cs(41,4): error RS0016: Symbol 'Command.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net9.0-android]
D:\a\_work\1\s\src\Controls\src\Core\CheckBox\CheckBox.cs(42,4): error RS0016: Symbol 'Command.set' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net9.0-android]
D:\a\_work\1\s\src\Controls\src\Core\CheckBox\CheckBox.cs(49,4): error RS0016: Symbol 'CommandParameter.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net9.0-android]
D:\a\_work\1\s\src\Controls\src\Core\CheckBox\CheckBox.cs(50,4): error RS0016: Symbol 'CommandParameter.set' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net9.0-android]
D:\a\_work\1\s\src\Controls\src\Core\CheckBox\CheckBox.cs(115,4): error RS0016: Symbol 'IsEnabledCore.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net9.0-android]
    56 Error(s)
```

rmarinho replied:
Can we add inline docs also?

rmarinho commented on src/Controls/src/Core/CheckBox/CheckBox.cs:
```diff
@@ -16,11 +18,40 @@ public partial class CheckBox : View, IElementConfiguration<CheckBox>, IBorderEl
+
+					checkBox.Handler?.UpdateValue(nameof(ICheckBox.Foreground));
+					checkBox.CheckedChanged?.Invoke(bindable, new CheckedChangedEventArgs((bool)newValue));
+					if (checkBox.Command is not null && checkBox.Command.CanExecute(null))
```
Should we pass the parameter to CanExecute?

rmarinho reviewed (changes requested):
Could we also add a simple UITest to show how it renderers check and uncheck and a Visual test. 

Thanks

pictos commented:
@rmarinho does maui use inline docs normally or there's some xml file that I should add?

pictos committed:
add public APIs to txt file

# Conflicts:
#	src/Controls/src/Core/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt

pictos committed:
pass CommandParameter to CanExecute method

pictos committed:
moved checkbox tests to the correct file

pictos commented:
> Could we also add a simple UITest to show how it renderers check and uncheck and a Visual test.

@rmarinho can you point me to a UITest that I can use as sample... I tried to look into the existing one on Appium project, but couldn't figure out how to create the UI to test

rmarinho commented:
> @rmarinho does maui use inline docs normally or there's some xml file that I should add?

We use inline docs now. @jfversluis  can help you with that if you need. 



> > Could we also add a simple UITest to show how it renderers check and uncheck and a Visual test.
> 
> @rmarinho can you point me to a UITest that I can use as sample... I tried to look into the existing one on Appium project, but couldn't figure out how to create the UI to test

https://github.com/dotnet/maui/blob/main/src/Controls/tests/UITests/Tests/Issues/Issue19379.cs#L27

jfversluis commented:
> does maui use inline docs normally or there's some xml file that I should add?

Yes please add the documentation inline and NOT the external XML files :)

mattleibow committed:
Update CheckBoxUnitTests.cs

mattleibow committed:
Merge branch 'net9.0' into pj/checkbox-command

mattleibow committed:
Update CheckBox.cs

mattleibow commented on src/Controls/src/Core/CheckBox/CheckBox.cs:
```diff
@@ -16,11 +18,38 @@ public partial class CheckBox : View, IElementConfiguration<CheckBox>, IBorderEl
+					}
+
+					checkBox.Handler?.UpdateValue(nameof(ICheckBox.Foreground));
+					checkBox.CheckedChanged?.Invoke(bindable, new CheckedChangedEventArgs((bool)newValue));
+					if (checkBox.Command is not null && checkBox.Command.CanExecute(checkBox.CommandParameter))
+					{
+						checkBox.Command.Execute(checkBox.CommandParameter);
+					}
```
Not sure what we should be doing here as we are going to fire the `CheckChanged` event but NOT the command if the command disabled it.

The command can disable the checkbox, and that is fine. But, if the dev then manually sets checked and we only fire one, will that be unexpected? Do we have a place where we have an event and command and we have to do something like this?

I found Button does this:

```cs
public static void ElementClicked(VisualElement visualElement, IButtonElement ButtonElementManager)
{
	if (visualElement.IsEnabled == true)
	{
		ButtonElementManager.Command?.Execute(ButtonElementManager.CommandParameter);
		ButtonElementManager.PropagateUpClicked(); // <- this is the clicked event
	}
}
```

mattleibow reviewed (changes requested):
The is looking close, but I had some questions/comments.

* Should the command and the event always fire (or not) together?
* Please add inline XML docs
* Please add some unit tests for the checkbox in the areas:
   * CheckChanged event
   * Command/CommandParameter
   * IsEnabled
* Please add some UI tests for checkbox:
   * Appears correctly in the correct state with Command CanExecute
   * Tapping the check box when enabled and disabled does the right thing

mattleibow commented:
/azp run

rmarinho commented:
/rebase

rmarinho commented:
Needs manual rebase

rmarinho commented:
Please reopen targeting the net10 branch
